### DISCUSSION TITLE
Source Generator Intellisense

### DISCUSSION THREAD

**Discussion by bforl** [dotnet/roslyn#75019] [Q&A]
I have 3 projects in my solution

- A Source Generator project that generates classes
- A library project that uses the above source generator
- A console app that references the library

For some reason, intelli-sense for the source generated classes only works inside the library project?
Intelli-sense for source generated classes does not work for the console app. I can write code with them, and it will compile, but I get no ide support for them

Is this to be expected?

**Comment by huoyaoyuan** [comment]
When you made any change with source generator, you need to do a full build then restart IDE, to ensure the latest generator is loaded in IDE.
Do you still observe the IntelliSense issue after restarting?

**Reply by bforl** [reply]
I can confirm that the IntelliSense issue is resolved after I restart the IDE, but how come I don't see the same issue with other 3rd part source generators that I use? For example, I use the Microsoft MVVM community toolkit for source generated `INotifyPropertyChanged`, and the IntelliSense updates fine without having to restart?

A little more info about my usage. 

I am trying to use a Source Generator in conjunction with Localisation strings to generate compile time constants for each localisation string key. To avoid defining strings in code that might end up failing at runtime.

- [SourceGenerator csproj] - Creates a 'Keys' class with static const strings by processing the AdditionalFiles from the context
- [Localisation csproj] - References the SourceGenerator proj, and contains an XML file of localisation key/value pairs which is configured as a AdditionalFiles entry in the csproj to be used by SourceGenerator 
- [ConsoleApp csproj] - References the Localisation proj to use the Keys class


The idea is that inside the console app, you should be able to use the generated `Keys` class to access Localisation keys in a compile time statically typed way. As there are so many keys, IntelliSense is quite important. If you have to restart the IDE every time you add a new entry to the XML to see it in IntelliSense, its not very helpful :(

Am I taking the wrong approach to solve the problem do you think?

**Reply by huoyaoyuan** [reply]
> but how come I don't see the same issue with other 3rd part source generators that I use?

Source generators shipped by packages are already compiled, and their path on disk will change if package versions are changed. IDE will always see an invariant, compiled assembly available.

> If you have to restart the IDE every time you add a new entry to the XML to see it in IntelliSense, its not very helpful :(

You should not be required to restart the IDE if you update the *data* used by the generator. Restart is only required when the *code* of generator changes.

> Am I taking the wrong approach to solve the problem do you think?

Not wrong, but you are somehow replicating the localization experience with `resx`.

**Reply by bforl** [reply]
> You should not be required to restart the IDE if you update the data used by the generator. Restart is only required when the code of generator changes.

I'm only updating the data. For example, all I do is go into the .xml and add a new line. At which point, when I go to the Cons app project, IntelliSense stops working for the Key class.


Would it help if I create a git hub repo showing the code?

**Reply by huoyaoyuan** [reply]
There are other possible reasons, for example if your generator doesn't handle subsequent or incremental calls gracefully. I'm not really sure about which code is responsible for this.

You can try to follow the latest source generator practices in https://github.com/dotnet/runtime/tree/main/src/libraries/System.Runtime.InteropServices/gen and see if anything improves.

**Reply by CyrusNajmabadi** [reply]
Intellisense is best effort with SGs.  Because of our extremely tight perf requirements, it is often the case that generators haven't completed by the time semantic information is needed by a feature like completion. 

We've also just had bugs in this area. Like not realizing a save of a file should rerun generators.

**Reply by bforl** [reply]
I've now tried moving the source generator to a completely different solution, building it once, and using it by **dll reference**. But I am getting the same issue. After changing the xml files in the Library (used by the source generator) the intellisense for the source generated code in the Lib project works, but it doesn't work in the console project :(

VS Solution 1
Source Generator

VS Solution 2
Library (references the dll by path from VS Solution 1)
Console (references the library by project reference)

I reference the analyser in the Library using the following : 
```
<ItemGroup>
	<Analyzer Include="C:\Source\KeysSourceGenerator\KeysSourceGenerator\bin\Debug\netstandard2.0\KeysSourceGenerator.dll" />
</ItemGroup>
```

**Reply by bforl** [reply]
I've just noticed that it seems to work much better with VS Professional? ...previously I was using the community edition.

**Reply by jasonmalinowski** [reply]
> I've just noticed that it seems to work much better with VS Professional? ...previously I was using the community edition.

Professional vs. Community edition won't matter here -- we don't change anything in this based on which edition you're using. Something else is causing that difference.

**Reply by bforl** [reply]
Hmm - ok. Not sure then. But things definitely started working better.
